Phase 1 Pixel DQM (number 3, final?)

schneiml opened the pull request:
This is a follow up to #14015 and #13798 .

All commits are rebased to the CMSSW_8_1_CLANG_X_2016-04-24-0000 IB such that #13987 lies below, with fixes to make things actually work.

Some points mentioned in #14015 are still relevant, but everything that I got feedback to is fixed.

@makortel regarding the Certification module, I now remove that in DQMOffline_CRT_cff.py, even though I still don't get how that finds its way into the offline config (cert-stuff is removed in SiPixelOfflineDQM). And, said module crashes due to code like this `float dcsFrac = (iGetter.get("Pixel/EventInfo/DCSSummary"))->getFloatValue();`, though I can't see immediatly which histogram is missing due to my changes (probably all...).

@slava77 I now make the TrackerTopology internal structs visible to the outside. I am not sure whether that is a good idea, but I agree it is better than what we had before.

@smuzaffar there are now 3 new packages, one with shared code and two that use this code for digis and clusters. More packages will come in the future, but I can't tell yet how we will call them.
UPDATE: now 4. Two for shared code and config, two for the plugins.

Overall, I think this is in a mergeable shape now. Let me know if you think different. Also, chasing IBs and rebasing ~100 commits each time starts to get annoying.

schneiml committed:
A from-scratch DQM Module

schneiml committed:
Minimal DQM config

schneiml committed:
Switch to sim digis; a thing to b centralized...

schneiml committed:
A bit of tracing output

schneiml committed:
PixelDQM framework skeleton. The hard stuff is not implemented yet, but the skeleton compiles.

schneiml committed:
FIx buildfile

schneiml committed:
TopFolder option for manager

schneiml committed:
Use Manager for Digis

schneiml committed:
Specify the SummationSpecification more precisely

schneiml committed:
A test harness that does not do the full DQM sequence

schneiml committed:
Simple booking. Not optimal yet, hardcoded ranges etc... but generates a plot\!

schneiml committed:
The actual test harness (file was missing in prev. commit)

schneiml committed:
WIP: read structure from alignment

schneiml committed:
WIP: Bitmap Extraction. Results for Pixel are still implausible

schneiml committed:
Some hierarchies of histograms. The Geometry seems implausible, needs to be checked.

schneiml committed:
Slightly modfied bitmasks.

Now the Geometry looks more sane but it does not resolve the Half* structures anymore; however I can't see manually how a bitmask could tell this.

schneiml committed:
Split into step1/step2 using a template skeleton.

This is less elegant than I hoped, but in return it gives us the power to remove most duplicated code. Note the ugly workaround needed for consumes<...>.

schneiml committed:
Allow for multiple histograms in one (Base-)Plugin

schneiml committed:
Make the interfaces a bit nicer

schneiml committed:
Implement more spec steps, and move the Builder impl out of the header.

schneiml committed:
Move module list to GeometryInterface, fix label problem

schneiml committed:
First working step2 histograms

schneiml committed:
Make ranges setable

schneiml committed:
Working EXTEND_X summation.

schneiml committed:
Improve metadata tracking. Now everything is a TH1D.

schneiml committed:
Switch to proper logging

schneiml committed:
Configure the message logger.

The warning level is spammed by cabling, but supressing seems to cause
a crash. Also LogDebug behaves differently from the higher ones, so I
stick to Info for some debug output.
(REBASE) The harvesting config file is added here since the earlier
commit was dropped.

schneiml committed:
Implement step1 step execution for the basic cases, with per-bx counts as an example. Lots of the new stuff are not well tested yet.

schneiml committed:
Implement string interening and or-columns.

Interning can be considered a premature optimization, but throwing around std::string in the hottest places is more of a design problem...

schneiml committed:
Use or-columns

schneiml committed:
Implement more 2D stuff. step2 EXTEND_Y is still missing.

schneiml committed:
Add columns for row/col. We need a way to have geometry-dependent parameters there.

schneiml committed:
Use the new 2D features for some hitmaps.

schneiml committed:
Turn the GeometryInterface into an edm::Service

schneiml committed:
Era-Based config for the common cases.

Many places are not updated yet. phase1TkCustoms is no longer used,
instead the DQM s re-enabled in DQMOffline_cff.py.

Conflicts:
	SLHCUpgradeSimulations/Configuration/python/phase1TkCustoms.py

schneiml committed:
add proper configuration for the HistogrgamManagers.

Ad-Hoc specs NYI. defaultGrouping() introduced here, though the interface is suboptimal.

schneiml committed:
a number of geometry improvements, including suppressing the UNDEFINED histogram frames.

schneiml committed:
Use consumesCollector; much cleaner than the old solution

schneiml committed:
Clean up the HistogramManager a bit (untested)

schneiml committed:
Fix broken spec

schneiml committed:
Refactoring the HistogramManager

schneiml committed:
New Python-Based spec builder

schneiml committed:
Port Digis Specs to Python

schneiml committed:
Read and actually use the Python specs

schneiml committed:
Move the rest of the Histogram config to Python

schneiml committed:
Remove the old test harness. With specifications in Python it is not really useful any more.

schneiml committed:
Remove now unneded configuration functionality

schneiml committed:
Remove the templated analyzer, specifications are now shared via config.

schneiml committed:
Fix bug in Hitmap spec

schneiml committed:
Obtain geometry data from TrackerTopology instead of Alignment

schneiml committed:
Also update configuration with new names

schneiml committed:
Instantiate one GeometryInterface per HistogramManager.

This avoids the trouble with edm::Service or a Singleton, but it may cause a problem in the future, since we cannot compare Columns of different HMs, since the interning tables might be different.

schneiml committed:
Disable two debug cout's

schneiml committed:
Make clang more happy

schneiml committed:
Fix uninitialized value causing a hard-to-reproduce crash

schneiml committed:
Make the GeometryInterface per-plugin. This sounds like a good compromise.

schneiml committed:
Allow for custom Harvesting (not used yet)

schneiml committed:
Allow ndigi-like specs to be executed. This involves special handling of per-Event things.

schneiml committed:
Make Specification Syntax a bit nicer

schneiml committed:
Fix Label in online-harvesting case

schneiml committed:
Add FEDs to the GeometryInterface

schneiml committed:
WIP: Make the Digis Plugin output the actual digis plots

schneiml committed:
Introduce a third stage (STAGE1_2) for per-event harvesting

schneiml committed:
Some fixed for the Digis plugin

schneiml committed:
Track max and min Values in GeometryInterface

schneiml committed:
Module is useless on endcap, use Panel instead

schneiml committed:
Split NDigis plots in two, to use different ranges

schneiml committed:
Add definitions of HalfCylinders and Shells to the GeometryInterface. These are not read from any central source.

Counting of Geometry Sizes (number of ladders etc.) is also added here, since it is necessary to interpret the ladder/blade counts to find the Shells.

schneiml committed:
Fix Shells by tracking a per-layer max ladder.

schneiml committed:
Add some debugging plots.

schneiml committed:
Add a README that explains the framework

schneiml committed:
Random CHanges. Includes merging the Endcap folders.

schneiml committed:
Add roc-level structure

schneiml committed:
More documentation in the README

schneiml committed:
Fix endcap rings.

schneiml committed:
Clean up histogram path construction, allow special cases there.

schneiml committed:
Extend README

schneiml committed:
Enable per-ring grouping for the endcaps.

schneiml committed:
Allow per-spec enable, and use it for PerModule switching

(REBASE) This also fixes a config problem due to dropped config commits.
Older commits probably don't run due to broken config.

schneiml committed:
Convert config to the style advertised in the readme

(named objects per quantity)

schneiml committed:
Reflect new per-module switch in the README

schneiml committed:
Add a Phase1 Cluster plugin.

schneiml committed:
Enable the new Plugin

The Harvesting config now includes the Source config, which is good to keep consistence but might cause other problems.

schneiml committed:
Sanitize bookUndefined behaviour

schneiml committed:
Reduce number of allocations in step1

schneiml committed:
Make Disk number signed

schneiml committed:
Remove a useless copy that caused loads of unnecessary allocations.

schneiml committed:
PERF: Cache last extracted Column Values. Reduces calls into the GeometryInterface by about 50%.

schneiml committed:
PERF: Reduce the number of map lookups in step1. Some of these optimisations are a bit risky.

schneiml committed:
Allow proper 2D ranges

schneiml committed:
Move geometry parameters to config.

HalfCylider/Shell should als be correct now.

schneiml committed:
Make folder names prettier.

schneiml committed:
Move naming stuff to GeometryInterface as Slava suggested.

This is also cleaner since some names have to be defined there anyways.

schneiml committed:
Disable SiPixelCertification, since it causes crashes.

I still don't see how this plugin gets into the SiPixelOffline config, but somehow it does and removing it here fixes the problem.

cmsbuild commented:
A new Pull Request was created by  @schneiml (Marcel Schneider) for CMSSW_8_1_X.

It involves the following packages:

DQM/SiPixelCommon
DQM/SiPixelPhase1Clusters
DQM/SiPixelPhase1Common
DQM/SiPixelPhase1Digis
DQMOffline/Configuration
DataFormats/TrackerCommon


The following packages do not have a category, yet:

DQM/SiPixelPhase1Clusters
DQM/SiPixelPhase1Common
DQM/SiPixelPhase1Digis

@cvuosalo, @dmitrijus, @cmsbuild, @deguio, @slava77, @vanbesien, @davidlange6 can you please review it and eventually sign? Thanks.
@makortel, @gpetruc, @rociovilar, @threus, @VinInn this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 9b4ab1203497b39163488bc4785ce62838005a21

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14230/12618/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
8.0 step1
<pre>runTheMatrix-results/8.0_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS/step1_BeamHalo+BeamHalo+DIGICOS+RECOCOS+ALCABH+HARVESTCOS.log

</pre>9.0 step1
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step1_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>25.0 step1
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step1_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>1306.0 step1
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step1_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>1330.0 step1
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step1_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>4.22 step2
<pre>runTheMatrix-results/4.22_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC/step2_RunCosmics2011A+RunCosmics2011A+RECOCOSD+ALCACOSD+SKIMCOSD+HARVESTDC.log

</pre>25202.0 step1
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step1_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>101.0 step1
<pre>runTheMatrix-results/101.0_SingleElectronE120EHCAL+SingleElectronE120EHCAL/step1_SingleElectronE120EHCAL+SingleElectronE120EHCAL.log

</pre>50202.0 step1
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step1_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

</pre>5.1 step1
<pre>runTheMatrix-results/5.1_TTbar+TTbarFS+HARVESTFS/step1_TTbar+TTbarFS+HARVESTFS.log

</pre>134.911 step2
<pre>runTheMatrix-results/134.911_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_2016+RECODR2_2016reHLT+HARVESTDR2/step2_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_2016+RECODR2_2016reHLT+HARVESTDR2.log

</pre>140.53 step2
<pre>runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/step2_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI.log

</pre>10021.0 step1
<pre>runTheMatrix-results/10021.0_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017/step1_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017.log

</pre>10024.0 step1
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017/step1_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017.log

</pre>1000.0 step2
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step2_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>1001.0 step2
<pre>runTheMatrix-results/1001.0_RunMinBias2011A+RunMinBias2011A+TIER0EXP+ALCAEXP+ALCAHARVD1+ALCAHARVD2+ALCAHARVD3+ALCAHARVD4/step2_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM.log

</pre>1003.0 step2
<pre>runTheMatrix-results/1003.0_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM/step2_RunMinBias2012A+RunMinBias2012A+RECODDQM+HARVESTDDQM.log

</pre>

schneiml commented:
I cannot reproduce these problems after a clean

    cmsrel CMSSW_8_1_X_2016-04-24-0000
    cd CMSSW_8_1_X_2016-04-24-0000/src
    cmsenv
    git cms-merge-topic 14230
    scram b -j4 # takes forever since TrackerTopology is touched
    runTheMatrix.py -l 1325.0 # etc.

I will however push soon a fix for the static analyzer warnings.

Judging by the name `frontierHttpClnt_usinglastproxyingroup` where the tests crash, I'd guess it is a problem on the build machine. (Also, I wonder where the source code for that lives?)

schneiml committed:
Fix static analyzer warnings (valid though not critical) and remove a useÃless import in TrackerTopology.

cmsbuild commented:
Pull request #14230 was updated. @cvuosalo, @dmitrijus, @cmsbuild, @deguio, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

deguio commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 29fc5cebd3def1bd14e5f6794aaad5f074a7ed23
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14230/12637/summary.html

deguio commented:
as discussed at todays meeting, the usage of th1f is suggested in place of the th1d histos
apart from that the PR looks good and can go in. more development will come on top for the online DQM.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14230/12637/summary.html

schneiml committed:
Remove some edm::Service traces

schneiml committed:
Switch to TH1F in harvesting, and other small cleanups

slava77 commented on DQM/SiPixelPhase1Common/python/MessageLogger_cfi.py:
```diff
@@ -0,0 +1,33 @@
+import FWCore.ParameterSet.Config as cms
+
+MessageLogger = cms.Service(
```
this overrides the central definition of MessageLogger when included in standard configuration.
This should not be done.

slava77 commented:
There is a new warning showing up at beginrun, apparently in all workflows (run1, run2, 2017)
```
%MSG-w SiStripCabling:   SiStripMonitorCluster:SiStripCalZeroBiasMonitorCluster@streamBeginRun  26-Apr-2016 15:48:16 CEST Run: 165121
[FedChannelConnection::apvPairNumber] Unexpected number of APV pairs: 65535
```
this has to be fixed (use eras to configure correctly) or silenced.

slava77 commented on DataFormats/TrackerCommon/interface/TrackerTopology.h:
```diff
@@ -576,7 +576,10 @@ class TrackerTopology {
-
+ 
+  PixelBarrelValues const& getPBVals() const { return pbVals_; };
+  PixelEndcapValues const& getPFVals() const { return pfVals_; };
```
since you need to update in other files, please remove  semicolons at the end of the two lines above

schneiml committed:
Remove the custom MessageLogger.

The debug log is not very useful and a performance problem anyways. This also suppresses the 'Unexpected number of APV pairs' warning.

schneiml committed:
Remove the Common (Code) -> Plugin -> Common (Config) dependency cycle.

This introduces a separate package for the config, which is not yet
common in Tracker DQM but seems like a good idea.

cmsbuild commented:
Pull request #14230 was updated. @cvuosalo, @dmitrijus, @cmsbuild, @deguio, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
the last test failed with ``` socket.timeout: timed out ```
and the currently available IB is broken

cmsbuild commented:
-1

Tested at: 3c4593964ff3a5935445405996d7c7cf3f67fd25

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14230/12701/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
134.911 step2
<pre>runTheMatrix-results/134.911_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_2016+RECODR2_2016reHLT+HARVESTDR2/step2_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_2016+RECODR2_2016reHLT+HARVESTDR2.log

</pre>1306.0 step2
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step2_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>1330.0 step2
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step2_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>10021.0 step2
<pre>runTheMatrix-results/10021.0_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017/step2_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017.log

</pre>25202.0 step2
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>10024.0 step2
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_PRef --relval 9000,50 --datatier "RAW" --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --eventcontent RAW --fileout file:RelVal_Raw_PRef_DATA.root --filein /store/data/Run2015D/MuonEG/RAW/v1/000/256/677/00000/80950A90-745D-E511-92FD-02163E011C5D.root : FAILED - time: date Thu Apr 28 22:32:04 2016-date Thu Apr 28 22:27:27 2016 s - exit: 16896
cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=HeavyIons -n 10 --conditions auto:run2_hlt_HIon --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016,Run2_HI --magField 38T_PostLS1 --fileout file:RelVal_Raw_HIon_DATA.root --filein /store/hidata/HIRun2015/HIHardProbes/RAW-RECO/HighPtJet-PromptReco-v1/000/263/689/00000/1802CD9A-DDB8-E511-9CF9-02163E0138CA.root : FAILED - time: date Thu Apr 28 22:32:02 2016-date Thu Apr 28 22:27:32 2016 s - exit: 16896
cmsDriver.py TTbar_Tauola_13TeV_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_GRun --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --fileout file:RelVal_Raw_GRun_MC.root : FAILED - time: date Thu Apr 28 22:35:43 2016-date Thu Apr 28 22:27:33 2016 s - exit: 16896
cmsDriver.py TTbar_Tauola_13TeV_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_PRef --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --fileout file:RelVal_Raw_PRef_MC.root : FAILED - time: date Thu Apr 28 22:34:39 2016-date Thu Apr 28 22:27:40 2016 s - exit: 16896
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,HLT:@relval2016,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 --magField 38T_PostLS1 : FAILED - time: date Thu Apr 28 22:36:18 2016-date Thu Apr 28 22:27:41 2016 s - exit: 16896
cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_PIon --relval 9000,50 --datatier "RAW" --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --eventcontent RAW --fileout file:RelVal_Raw_PIon_DATA.root --filein /store/data/Run2015D/MuonEG/RAW/v1/000/256/677/00000/80950A90-745D-E511-92FD-02163E011C5D.root : FAILED - time: date Thu Apr 28 22:32:11 2016-date Thu Apr 28 22:28:24 2016 s - exit: 16896
cmsDriver.py TTbar_Tauola_13TeV_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=HeavyIons -n 10 --conditions auto:run2_mc_HIon --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016,Run2_HI --magField 38T_PostLS1 --fileout file:RelVal_Raw_HIon_MC.root : FAILED - time: date Thu Apr 28 22:38:17 2016-date Thu Apr 28 22:32:10 2016 s - exit: 16896
cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_GRun --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --fileout file:RelVal_Raw_GRun_DATA.root --filein /store/data/Run2015D/MuonEG/RAW/v1/000/256/677/00000/80950A90-745D-E511-92FD-02163E011C5D.root : FAILED - time: date Thu Apr 28 22:40:21 2016-date Thu Apr 28 22:34:53 2016 s - exit: 16896
cmsDriver.py TTbar_Tauola_13TeV_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_PIon --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --fileout file:RelVal_Raw_PIon_MC.root : FAILED - time: date Thu Apr 28 22:42:05 2016-date Thu Apr 28 22:36:09 2016 s - exit: 16896

cmsbuild commented:
-1

Tested at: 3c4593964ff3a5935445405996d7c7cf3f67fd25

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14230/12707/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
134.911 step2
<pre>runTheMatrix-results/134.911_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_2016+RECODR2_2016reHLT+HARVESTDR2/step2_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_2016+RECODR2_2016reHLT+HARVESTDR2.log

</pre>1306.0 step2
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step2_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>1330.0 step2
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step2_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>10021.0 step2
<pre>runTheMatrix-results/10021.0_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017/step2_TenMuE_0_200+TenMuE_0_200_pythia8_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017.log

</pre>25202.0 step2
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>10024.0 step2
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_PRef --relval 9000,50 --datatier "RAW" --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --eventcontent RAW --fileout file:RelVal_Raw_PRef_DATA.root --filein /store/data/Run2015D/MuonEG/RAW/v1/000/256/677/00000/80950A90-745D-E511-92FD-02163E011C5D.root : FAILED - time: date Fri Apr 29 02:17:50 2016-date Fri Apr 29 02:11:57 2016 s - exit: 16896
cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=HeavyIons -n 10 --conditions auto:run2_hlt_HIon --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016,Run2_HI --magField 38T_PostLS1 --fileout file:RelVal_Raw_HIon_DATA.root --filein /store/hidata/HIRun2015/HIHardProbes/RAW-RECO/HighPtJet-PromptReco-v1/000/263/689/00000/1802CD9A-DDB8-E511-9CF9-02163E0138CA.root : FAILED - time: date Fri Apr 29 02:17:17 2016-date Fri Apr 29 02:12:01 2016 s - exit: 16896
cmsDriver.py TTbar_Tauola_13TeV_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_GRun --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --fileout file:RelVal_Raw_GRun_MC.root : FAILED - time: date Fri Apr 29 02:22:31 2016-date Fri Apr 29 02:12:04 2016 s - exit: 16896
cmsDriver.py TTbar_Tauola_13TeV_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_PRef --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --fileout file:RelVal_Raw_PRef_MC.root : FAILED - time: date Fri Apr 29 02:20:59 2016-date Fri Apr 29 02:12:12 2016 s - exit: 16896
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,HLT:@relval2016,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 --magField 38T_PostLS1 : FAILED - time: date Fri Apr 29 02:24:16 2016-date Fri Apr 29 02:12:13 2016 s - exit: 16896
cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_PIon --relval 9000,50 --datatier "RAW" --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --eventcontent RAW --fileout file:RelVal_Raw_PIon_DATA.root --filein /store/data/Run2015D/MuonEG/RAW/v1/000/256/677/00000/80950A90-745D-E511-92FD-02163E011C5D.root : FAILED - time: date Fri Apr 29 02:19:15 2016-date Fri Apr 29 02:13:17 2016 s - exit: 16896
cmsDriver.py TTbar_Tauola_13TeV_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=HeavyIons -n 10 --conditions auto:run2_mc_HIon --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016,Run2_HI --magField 38T_PostLS1 --fileout file:RelVal_Raw_HIon_MC.root : FAILED - time: date Fri Apr 29 02:26:51 2016-date Fri Apr 29 02:17:53 2016 s - exit: 16896
cmsDriver.py RelVal -s L1REPACK:Full --data --scenario=pp -n 10 --conditions auto:run2_hlt_GRun --relval 9000,50 --datatier "RAW" --eventcontent RAW --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --fileout file:RelVal_Raw_GRun_DATA.root --filein /store/data/Run2015D/MuonEG/RAW/v1/000/256/677/00000/80950A90-745D-E511-92FD-02163E011C5D.root : FAILED - time: date Fri Apr 29 02:29:23 2016-date Fri Apr 29 02:21:24 2016 s - exit: 16896
cmsDriver.py TTbar_Tauola_13TeV_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW --mc --scenario=pp -n 10 --conditions auto:run2_mc_PIon --relval 9000,50 --datatier "GEN-SIM-RAW" --eventcontent RAWSIM --customise=HLTrigger/Configuration/CustomConfigs.L1T --era Run2_2016 --magField 38T_PostLS1 --fileout file:RelVal_Raw_PIon_MC.root : FAILED - time: date Fri Apr 29 02:31:14 2016-date Fri Apr 29 02:23:10 2016 s - exit: 16896

slava77 commented:
@cmsbuild please test
... apparently, tests that were started more recently now complete

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 3c4593964ff3a5935445405996d7c7cf3f67fd25
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14230/12731/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14230/12731/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/14230	3c45939
- changes in reco are rather minimal, exposing internal structure of TrackerTopology in PixelBarrelValues and PixelEndcapValues, needed for DQM organized partly using this internal layout.
- jenkins tests pass and comparisons with baseline show no difference (compared to the last checked commits as of 29fc5ce, the log files are now clean)

dmitrijus commented:
+1

schneiml commented:
@davidlange6 can we get the new packages added and this PR merged?

fioriNTU commented:
@davidlange6 @slava77 the Phase I DQM is a fundamental tool for us, we need to test it extensively in the Tracker DQM group, use it to validate the new geometry and relvals. It would be very important to have this PR merged as soon as possible. It would also ease the share of work in our group. Do you have a time estimate to have it done, and possibly to have a prerelease with this code?

fioriNTU commented:
@davidlange6 @slava77  Sorry to insist, but this is really crucial for us. Any news about it?

slava77 commented:
@fioriNTU 
I signed this PR a while back (2 weeks ago).
Nominally, all development release PRs (81X) get merged within a day or two after they are fully signed,
unless there are some issues.

@davidlange6 are there any issues with the changes made here?

davidlange6 commented:
yes, I don’t like the change to TrackerTopology.h as I have explained to @fioriNTU (exposing the internal implementation of the DetId packing). I have yet to trace through why it is needed for this PR.


> On May 17, 2016, at 3:16 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> @fioriNTU 
> I signed this PR a while back (2 weeks ago).
> Nominally, all development release PRs (81X) get merged within a day or two after they are fully signed,
> unless there are some issues.
> 
> @davidlange6 are there any issues with the changes made here?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

slava77 commented:
On 5/17/16 6:23 AM, David Lange wrote:
> yes, I don’t like the change to TrackerTopology.h as I have explained to
> @fioriNTU (exposing the internal implementation of the DetId packing). I
> have yet to trace through why it is needed for this PR.
> 

OK, good to know.
I suggested to expose this after seeing the original implementation,
which was string based and rather DQM-specialized solution for something
that looked like "give me all you know about pixel subdetectors".
Probably another solution with more sensible interface from the
TrackerTopology can work out better.
It better be enum-based with string formatting done on the client code side.

	

> 
>> On May 17, 2016, at 3:16 PM, Slava Krutelyov <notifications@github.com> wrote:
>> 
>> @fioriNTU 
>> I signed this PR a while back (2 weeks ago).
>> Nominally, all development release PRs (81X) get merged within a day or two after they are fully signed,
>> unless there are some issues.
>> 
>> @davidlange6 are there any issues with the changes made here?
>> 
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14230#issuecomment-219715795>
>

davidlange6 commented:
> On May 17, 2016, at 4:09 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 5/17/16 6:23 AM, David Lange wrote:
> > yes, I don’t like the change to TrackerTopology.h as I have explained to
> > @fioriNTU (exposing the internal implementation of the DetId packing). I
> > have yet to trace through why it is needed for this PR.
> > 
> 
> OK, good to know.
> I suggested to expose this after seeing the original implementation,
> which was string based and rather DQM-specialized solution for something
> that looked like "give me all you know about pixel subdetectors".
> Probably another solution with more sensible interface from the
> TrackerTopology can work out better.
> It better be enum-based with string formatting done on the client code side.

TrackerTopology doesn’t tell you anything about the pixel sub detector actually included in the CMS geometry - but rather some maximal properties (maximum number of layers). I don’t see why that is interesting. It is instead the actual geometry that is interesting to monitor. anyway, I still haven’t looked at an example...

> 
> 
> 
> > 
> >> On May 17, 2016, at 3:16 PM, Slava Krutelyov <notifications@github.com> wrote:
> >> 
> >> @fioriNTU 
> >> I signed this PR a while back (2 weeks ago).
> >> Nominally, all development release PRs (81X) get merged within a day or two after they are fully signed,
> >> unless there are some issues.
> >> 
> >> @davidlange6 are there any issues with the changes made here?
> >> 
> >> —
> >> You are receiving this because you were mentioned.
> >> Reply to this email directly or view it on GitHub
> >> 
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > <https://github.com/cms-sw/cmssw/pull/14230#issuecomment-219715795>
> > 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

davidlange6 commented:
After more investigation I don't really see how the mask bit offsets are used here. Can you summarize? Contrary to what @slava77  says, this question was not due to his suggestion, as the code was just moved from one spot to another. @schneiml @fioriNTU

schneiml commented:
The mask bits from TackerTopology are used to extract the respective qauntities from DetIds, as TT does. However, the interface to be used should not be based on method names, but allow
- Listing all quantities
- associating each quantity with a human-readable name
which is both not possible with TT. While the original plan was to provide TT with this functionality (which I'd still prefer), we moved the name-mapping to the PixelDQM-specific =GeometryInterface= since the user-facing names should be DQM-specific.

The information is in turn transformed into another representation (vector of functors), which is more specific and may contain bitmask-based as well as derived quantities.

davidlange6 commented:
what is a “name” in this context? can you give an example?
(and its not a geometry, so GeometryInterface should be renamed)

> On May 19, 2016, at 9:12 AM, Marcel Schneider <notifications@github.com> wrote:
> 
> The mask bits from TackerTopology are used to extract the respective qauntities from DetIds, as TT does. However, the interface to be used should not be based on method names, but allow
> 
> 	• Listing all quantities
> 	• associating each quantity with a human-readable name which is both not possible with TT. While the original plan was to provide TT with this functionality (which I'd still prefer), we moved the name-mapping to the PixelDQM-specific =GeometryInterface= since the user-facing names should be DQM-specific.
> The information is in turn transformed into another representation (vector of functors), which is more specific and may contain bitmask-based as well as derived quantities.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

schneiml commented:
Human-readable names are the strings visible around here: https://github.com/cms-sw/cmssw/pull/14230/files#diff-cf515bc8af8720e3dcf9d5d81ec7607fR201

These will later also be visible in the DQM, e.g. in the GUI. For this very reason I now think it is not useful to have them in TT, since we might want to change them later. However, I do think CMSSW might want to have a convention cast in code on how to label such things.

The `GeometryInterface` is meant to be an interface, as the name suggests, to all potentially geometry-dependent things. The fact that there is a lot of information about Topology encoded is due to the problem that there is no central place (to my knowledge) that has all the required information (esp. in an reasonably generic yet not too slow way). E.g. the Shell concept of DQM and the associated signed counting convention for ladders seems to be ad-hoc hardcoded in many places atm, so it has to be hardcoded in the `GeometryInterface` as well.

davidlange6 commented:
> On May 19, 2016, at 9:31 AM, Marcel Schneider <notifications@github.com> wrote:
> 
> Human-readable names are the strings visible around here: https://github.com/cms-sw/cmssw/pull/14230/files#diff-cf515bc8af8720e3dcf9d5d81ec7607fR201
> 

yes, I saw that, but then I didn’t manage to follow the actual usage - 

> These will later also be visible in the DQM, e.g. in the GUI. For this very reason I now think it is not useful to have them in TT, since we might want to change them later. However, I do think CMSSW might want to have a convention cast in code on how to label such things.
> 

pixels have some already, e.g.,
DataFormats/SiPixelDetId/interface/Pixel*Name*h


> The GeometryInterface is meant to be an interface, as the name suggests, to all potentially geometry-dependent things. The fact that there is a lot of information about Topology encoded is due to the problem that there is no central place (to my knowledge) that has all the required information (esp. in an reasonably generic yet not too slow way). E.g. the Shell concept of DQM and the associated signed counting convention for ladders seems to be ad-hoc hardcoded in many places atm, so it has to be hardcoded in the GeometryInterface as well.
> 

yes, its too bad that this is still the case as its not a new issue. There are some non-hard coded examples in pixels to follow such as DQM/SiPixelMonitorTrack/src/SiPixelTrackResidualSource.cc


> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

schneiml commented:
> pixels have some already, e.g.,
> DataFormats/SiPixelDetId/interface/Pixel*Name*h

Except that there are no actual "names" in there. There is some string
parsing, but I fail to see the corresponding formatting. Even if it were
available, this is not enough for the DQM, where we will have a folder
structure that contains some, but not all, of the fields.

> such as DQM/SiPixelMonitorTrack/src/SiPixelTrackResidualSource.cc

I don't know SiPixelTrackResidualSource by heart (yet), but I fail to
rememember almost anything *not* hardcoded there. I guess you are just
referring to the use of the classes mentioned above?

(looking more closely, `SiPixelTrackResidualSource.cc` is indeed not as
bad as other Pixel DQM plugins.)

davidlange6 commented:
@schneiml - what sort of string do you need? A human readable, physically meaningful, and unique name corresponding each DetId?

schneiml commented:
> @schneiml - what sort of string do you need? A human readable, physically meaningful, and unique name corresponding each DetId?

No, I really want to see the structure. The DQM _might_ want to see all
of the "field" names and values ("Ladder_4", "Module_1", "Shell_pI") but
likely we don't want all of them but rather a subset. (e.g. Most often
we don't care about the module).

This is what the `namedPartitions` map allows: it lists all available
"fields", including the name and a way to acces the value of this field
(the information whether the field is applicable to some given DetId is
also important and ends to be encoded nowhere).

So, human readable and physically meaningful, but not necessarily
unique, since we care a lot about groups of DetIds (like full ladders,
full layers, etc.)

davidlange6 commented:
Ok - So I don't think anyone can provide whats needed without some specifications from DQM in that case.  However, what you need to do is to use the TrackerGeometry to derive the ranges of these things rather than the topology. [best would be to have a ESSource that depends on the geometry which provides a central location of the needed info, but there seems to be not so much interest in that old idea]

schneiml commented:
Putting this stuff into an ESSource is a longer term goal, but we first have to figure out what we actually need (this is sort of understood now, but was not in the beginning) and also see what other users might be interested in. Until then we will have a local solution (maybe for ever). 

TrackerGeometry is used to figure out precise ranges, though that needs some reworking in the future - atm it e.g. does not represent the number of ladders _per layer_. But this does not matter for now, since this sort of information is (atm) only used in harvesting where ad-hoc counting is no problem. 

So the `GeometryInterface` will remain a facade around other things including TT, and the design space ranges from _encoding all relvant information there_ to having _only an interface_ to other sources (as originally intended). At the moment we have a compromise that uses the bitmasks from TT and counts from TG (and also Cabling) but encodes other things in place (Naming and Shells).

Are you ok with this solution?

davidlange6 commented:
@schneiml - I would find it reasonable to expose the maximum number of Pixel barrel (endcap, strip blah blah) layers (modules, other quantities, etc) in TrackerTopology via accessors returning unsigned ints, which sounds like it would be sufficient for your needs.

schneiml commented:
@davidlange6 the maximum number is not really the main problem, since TrackerGeometry already provides this (not in a well readable form, but good enough that `GeometryInterface` can figure it out).

The main question I see is whether 
- we want to avoid changing TrackerTopology (which means a more ugly facade for me, since I need to wrap the methods in TT using an array of lambdas or sth. such that I can call them by-index or by-string-name, instead of by-methodname)
- or we want to fix the problem properly, which means providing a proper interface to access the fields of DetIds without using the different methods of TT (and probably more, to make it useful).

The third option is to leave things as they are, which is somewhere in the middle.

davidlange6 commented:
> On May 19, 2016, at 3:01 PM, Marcel Schneider <notifications@github.com> wrote:
> 
> @davidlange6 the maximum number is not really the main problem, since TrackerGeometry already provides this (not in a well readable form, but good enough that GeometryInterface can figure it out).
> 
> The main question I see is whether
> 
> 	• we want to avoid changing TrackerTopology (which means a more ugly facade for me, since I need to wrap the methods in TT using an array of lambdas or sth. such that I can call them by-index or by-string-name, instead of by-methodname)
> 	• or we want to fix the problem properly, which means providing a proper interface to access the fields of DetIds without using the different methods of TT (and probably more, to make it useful).
> The third option is to leave things as they are, which is somewhere in the middle.
> 

I have no problem with a better interface to the existing exposed information. (but that is not what I’ve seen in the PR either before or after @slava77 comments


> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>

schneiml commented:
A better interface will take a considerable amount of extra work. The original idea of putting the human-readable names into TT as well turns out to be not really practicable. A variant with enum-keys might work, one might even find an appropriate enum somewhere in the code already. 

However, if we want to keep the bit masking all contained in TT (which seems like a good idea) and do not want to hand out functors (which sounds way to compilcated) we need to make sure that this is fast enough to be useful.

I could imagine a method 

     unsigned int getField(const DetId& id, SuitableEnum idx) { return ((id.rawId()>>startbit[idx])& mask[idx]);}

but this requires access to the bitmasks as an array. Something like this does not exist yet (or does it? serialization seems to have something in this direction but it sounds like we don't want to use that). 

This means a significant amount of planning and implementation.

davidlange6 commented:
@schneiml - I like the proposal, yes (possibly adding a sub detector argument as well).  adding a array of refs for each of the relevant fields should be straightforward

schneiml commented:
Ok, I can try to implement that, however

- This will mean a new PR. The IB of this one long dead.
- Lots of rebasing to get everything up to date. I'd then add the other ~20 commits I have accumulated as well (a lot more files, more packages, more histograms, but nothing spectacular in there).

I'd go that path if we can get the new PR merged next week.  With another 30 day delay this is unacceptable. 

If you have any other suggestions, objections, renaming wishes, whatever, please tell me now.

schneiml commented:
@davidlange6 I have now implemented something here:
[Data](https://github.com/schneiml/cmssw/blob/phase1pixeldqm-fourth/DataFormats/TrackerCommon/interface/TrackerTopology.h#L105) [Functions](https://github.com/schneiml/cmssw/blob/phase1pixeldqm-fourth/DataFormats/TrackerCommon/interface/TrackerTopology.h#L607)

Looks good to me, but causes ICE on GCC (4.9 as well as 5.1, clang accepts it), so we'll have to change it a bit (maybe moving the initializer to the .cc).

Please comment on whether this is what you expected.

davidlange6 commented:
@schneiml - yes, that looks good to me - i guess the pixel naming convention seems to be pf (or pxf) rather than pe for some reason.

schneiml commented:
New PR #14586 is there. This one can probably be closed, if there are no issues with the new one.

Regarding the Endcap naming -- this is farly inconsistent, I defined the
difference to be that there is 2 PEs but only one P(X)F. But that does
not really make a difference in the given case, so I'll change to PF, to
be more in line with the rest of TT.

davidlange6 commented:
I agree that PE would seem more consistent..

davidlange6 commented:
I'll close this in favor of v4 - which hopefully @dmitrijus has a look at in the next days
